arch-riscv: Add PCEvent for RISCV FS Workload kernel panic/oops

hnpl opened the pull request:
Inspired by the similar feature in ARM's full system workload, this change adds
an option to halt gem5 simulation if the guest system encounter kernel panic
or kernel oops.

On RiscvISA::BootloaderKernelWorkload, by default, the simulation
will exit upon kernel panic, while kernel oops will not induce simulation halt.
This is because the system will essentially do nop after a kernel panic, while the
system might be still functional after a kernel oops.

Dumping kernel's dmesg is useful for diagonizing the cause of kernel panic, so
ideally, we want to dump the guest's dmesg to the host. However, due to a bug
described in [1], kernel v5.18+ dmesg might not be dumped properly. Hence, the
dmesg will not be dumped to the host.

On RiscvISA::FsLinux, this feature is turned off by default as the symbols from the
official RISC-V kernel resource are stripped from the binary. However, if this feature
is enable, the dmesg will be dumped to the host system.

[1] https://github.com/gem5/gem5/issues/550

Change-Id: I8f52257727a3a789ebf99fdd4dffe5b3d89f1ebf

hnpl committed:
arch-riscv: Add PCEvent for RISCV FS Workload kernel panic/oops

Inspired by a similar feature in ARM's full system workload, this change adds
an option to halt gem5 simulation if the guest system encounter kernel panic
or kernel oops.

On RiscvISA::BootloaderKernelWorkload, by default, the simulation
will exit upon kernel panic, while kernel oops will not induce simulation halt.
This is because the system will essentially do nop after a kernel panic, while the
system might be still functional after a kernel oops.

Dumping kernel's dmesg is useful for diagonizing the cause of kernel panic, so
ideally, we want to dump the guest's dmesg to the host. However, due to a bug
described in [1], kernel v5.18+ dmesg might not be dumped properly. Hence, the
dmesg will not be dumped to the host.

On RiscvISA::FsLinux, this feature is turned off by default as the symbols from the
official RISC-V kernel resource are stripped from the binary. However, if this feature
is enable, the dmesg will be dumped to the host system.

[1] https://github.com/gem5/gem5/issues/550

Change-Id: I8f52257727a3a789ebf99fdd4dffe5b3d89f1ebf
Signed-off-by: Hoa Nguyen <hn@hnpl.org>
Co-authored-by: Jason Lowe-Power <jason@lowepower.com>

BobbyRBruce reviewed (dismissed):
Giving a soft-approval. I'd like someone with a bit more RISC-V Knowledge to have a look at it.

BobbyRBruce commented:
@rjc-arch @rogerchang23424 : Could one of you see if this looks right to you?

rogerchang23424 reviewed (commented):
It's OK to me.

powerjg commented on src/arch/riscv/linux/fs_workload.cc:
```diff
@@ -196,5 +260,11 @@ BootloaderKernelWorkload::unserialize(CheckpointIn &checkpoint)
+void
+Gem5Exit::process(ThreadContext *tc)
+{
+    fatal(descr());
```
I think this should be a `panic` since it's not an incorrect configuration. 

Really, I'm not sure if it should be a `panic` *or* a `fatal`. What is this purpose of this event?

hnpl replied:
I think the best way to describe this event is to exit gem5 when there's a kernel panic.

I leaned towards using `fatal` because `panic` implies there's something with gem5. A kernel panic could also be a product of a wrong configuration, e.g., using a wrong boot parameter.

Another reason is that `panic` dumps the backtrace while `fatal` does not. The backtrace might be misleading as the exit message is pushed further upward.

Anyway, if `fatal` means there __must__ be something wrong with the configuration then I'll switch to use `panic`.

powerjg replied:
If the goal is to exit the gem5 simulation loop, then you should call `exitSimLoop` (this isn't the exact name of the function, but it's something like that). I agree that neither `panic` nor `fatal` is right in this case.

rjc-arch replied:
+1 it's `exitSimLoop` from `sim/sim_exit.hh`.

hnpl replied:
Thanks for the suggestions. I updated the exit event accordingly.

I want to note that, for ArmFsWorkloads, guest's kernel panic will cause a gem5 panic, so we have different kernel panic exit behaviors for workloads of different ISAs. This might be confusing to users.

powerjg commented on src/arch/riscv/linux/fs_workload.hh:
```diff
@@ -44,11 +44,32 @@ namespace RiscvISA
     FsLinux(const Params &p) : KernelWorkload(p) {}
+    ~FsLinux()
+    {
+        if (kernelPanicPcEvent != nullptr)
+        {
+            delete kernelPanicPcEvent;
+        }
+        if (kernelOopsPcEvent != nullptr)
+        {
+            delete kernelOopsPcEvent;
+        }
```
This is not to style.

```suggestion
        if (kernelPanicPcEvent != nullptr) {
            delete kernelPanicPcEvent;
        }
        if (kernelOopsPcEvent != nullptr) {
            delete kernelOopsPcEvent;
        }
```

powerjg commented on src/arch/riscv/linux/fs_workload.hh:
```diff
@@ -87,7 +117,20 @@ class BootloaderKernelWorkload: public Workload
 
+    ~BootloaderKernelWorkload()
+    {
+        if (kernelPanicPcEvent != nullptr)
+        {
+            delete kernelPanicPcEvent;
+        }
+        if (kernelOopsPcEvent != nullptr)
+        {
+            delete kernelOopsPcEvent;
+        }
```
Also not to style.

```suggestion
        if (kernelPanicPcEvent != nullptr) {
            delete kernelPanicPcEvent;
        }
        if (kernelOopsPcEvent != nullptr) {
            delete kernelOopsPcEvent;
        }
```

rjc-arch reviewed (commented):
Hi @hnpl. The generic parts of this code look good as far as I can see.

hnpl committed:
stdlib: Add Kernel Panic/Oops exit event to stdlib

RISCV full system workloads have the capability of exit the simulation loop
upon the guest's kernel panic/oops. This change adds more stdlib exit event types
to accommodate the corresponding gem5 exits upon the guest's kernel panic and
kernel oops.

Change-Id: I3a4f313711793a473c6f138ff831b948034d0bb6
Signed-off-by: Hoa Nguyen <hn@hnpl.org>

powerjg commented on src/arch/riscv/RiscvFsWorkload.py:
```diff
@@ -84,5 +92,13 @@ class RiscvBootloaderKernelWorkload(Workload):
+
+    # gem5 event upon guest's kernel panic
+    exit_on_kernel_panic = Param.Bool(
+        True, "Generate gem5 exit upon the guest's kernel panic."
```
Why is this True when it is False in the RiscvFsWorkload.py

hnpl commented:
Current linux kernel from gem5 resources does not have symbol. This exit on
kernel panic behavior relies on being able to find the panic function
symbol. Default this to true on RiscvFsWorkload will crash the simulation.

On Thu, Nov 30, 2023, 10:15 Jason Lowe-Power ***@***.***>
wrote:

> ***@***.**** approved this pull request.
> ------------------------------
>
> In src/arch/riscv/RiscvFsWorkload.py
> <https://github.com/gem5/gem5/pull/573#discussion_r1411093453>:
>
> > @@ -84,5 +92,13 @@ class RiscvBootloaderKernelWorkload(Workload):
>
>      # booting parameters
>      boot_args = Param.String(
> -        "", "Booting arguments, to be passed to the kernel"
> +        "", "Booting arguments, to be passed to the kernel."
> +    )
> +
> +    # gem5 event upon guest's kernel panic
> +    exit_on_kernel_panic = Param.Bool(
> +        True, "Generate gem5 exit upon the guest's kernel panic."
>
> Why is this True when it is False in the RiscvFsWorkload.py
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gem5/gem5/pull/573#pullrequestreview-1758166175>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHSFRYNTXXJLTHFXEQTSDM3YHDENZAVCNFSM6AAAAAA7RN4KRCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONJYGE3DMMJXGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

powerjg commented:
Hmm. I think maybe we should always default to false and only set it to true if we know the resource has symbols. I'm not sure how to do that since the resource may not have that information easily available or have the API to make that update.

powerjg commented on src/arch/riscv/linux/fs_workload.cc:
```diff
@@ -75,6 +78,39 @@ FsLinux::initState()
+{
+    const std::string dmesg_output = name() + ".dmesg";
+    if (params().exit_on_kernel_panic) {
+        kernelPanicPcEvent = addFuncEventOrPanic<linux::KernelPanic>(
```
Use a different version of this function so that it doesn't panic if it can't find the symbol. Instead, we want to have a warning.

rjc-arch commented:
From a quick glance at the code it looks like we could use `addFuncEvent` instead of `addFuncEventOrPanic` to add the events, and then manually print a warning instead of panicking if `nullptr` is returned (in the case that the `panic`/`oops_exit` symbols are not found).

I'm making some changes to the Arm Panic/Oops handling at the moment related to #550; I've just tried making the same change there, and it seems to work. (I'm not sure if you want to wait to see what I have done before merging this change?)

(@powerjg You just got in ahead of me!)

powerjg commented:
I am going to take over this patch. Yeah, `addFuncEvent` is what I was thinking. When you push a patch for #550 I'll follow the same format.

@rjc-arch, do you think you'll have a patch for #550 by tomorrow?

rjc-arch commented:
I hope so. I'm just responding to some feedback from @andysan before pushing.

rjc-arch commented:
Hi @powerjg. I've just pushed #639 if you'd like to take a look. I'm not sure how well it fits in with what you were planning. I suspect we can hoist the exit events fully into the common code now that adding the events fails gracefully.

powerjg commented:
Yes, that looks great @rjc-arch! 

I think this change is going to need a bit more time, though. I need to understand why `BootloaderKernelWorkload` is not a subclass of `KernelWorkload`, which is preventing us from hoisting the duplicate code right now.

After #639 is merged, this should be rebased on top of that change (possibly with hoisting common code).

BobbyRBruce commented on src/arch/riscv/linux/fs_workload.hh:
```diff
@@ -87,7 +115,18 @@ class BootloaderKernelWorkload: public Workload
 
+    ~BootloaderKernelWorkload()
+    {
+        if (kernelPanicPcEvent != nullptr) {
+            delete kernelPanicPcEvent;
+        }
+        if (kernelOopsPcEvent != nullptr) {
+            delete kernelOopsPcEvent;
+        }
```
Fyi, you can safely "delete" `nullptr`. It's part of the CPP standard. It does nothing.

(Not a request for change. This is fine, just overly-verbose).

BobbyRBruce reviewed (commented):
I'm not sure what the intended affect is here but as currently written the exit events will do nothing by default (just continue the simulation loop). If you want a different default behavior (such as exit the simulation like on the `ExitEvent.EXIT`) you need to update this Dictionary: https://github.com/gem5/gem5/blob/569e21f798f61c22a91e8f6bad083a82fdf43034/src/python/gem5/simulate/simulator.py#L275C14-L275C35

powerjg commented:
Thanks, Bobby. I've updated the default dictionary.

powerjg commented:
@BobbyRBruce I have verified this works. Can you please re-review?
